### PR TITLE
chore: Enable G404 rule for gosec

### PR DESCRIPTION
**Pull request by zak-pawel**
PR:
- Introduces [github.com/chanxuehong/rand](github.com/chanxuehong/rand) dependency which provides random number/bytes related functions - it has interface very similar to `math/rand`, except that it prefers to use `crypto/rand` to implement functions.
- Enables `gosec:G404` rule.
- Fixes `gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)` for production code.
- Ignores mentioned findings for test code.

resolves: https://github.com/influxdata/telegraf/issues/12949

### CODE REVIEW

**Review by powersj — COMMENTED** [review]
> github.com/chanxuehong/rand

Why use this over using "crypto/rand" directly? I realize this is a small addition, but if we can do it without yet another dependency I would prefer that.

**Comment by zak-pawel** [comment]
> > github.com/chanxuehong/rand
> 
> Why use this over using "crypto/rand" directly? I realize this is a small addition, but if we can do it without yet another dependency I would prefer that.

@powersj 
This is API of [math/rand](https://pkg.go.dev/math/rand):
image

And this is API of [crypto/rand](https://pkg.go.dev/crypto/rand):
image

This library provides almost the same API as `math/rand` but prefers to use `crypto/rand`.

Alternatively we can write our own wrapper...

**Comment by zak-pawel** [comment]
> can you report how many of those functions we would need to wrap instead?

- `rand.Int31n`
- `rand.Int63`
- `rand.Int63n`
- `rand.Intn`
- `rand.Float64`

**Comment by zak-pawel** [comment]
@powersj Alternatively, we can just copy these 4-5 files to our repo, review correctness of algorithms (eg. compare with current implementation of `math/rand`) and have it in Telegraf's codebase.

**Comment by powersj** [comment]
Thanks for looking into how many functions we actually use.

With only 5 functions I would prefer we have our own copys in internal and not add the external dependency please. I don't think we should wholesale copy that project, but instead just write replacements for those 5 functions.

Thanks again

**Comment by powersj** [comment]
```
internal/internal.go:128:13                            gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)
internal/internal.go:146:13                            gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)
```

these two are for random sleep and random duration, probably not something that needs to be cryptographically secure, can ignore

```
plugins/inputs/example/example.go:106:13               gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)
```

example plugin, don't care about this other than maybe removing it from the example as a bad practice?

```
plugins/inputs/mock/mock.go:69:11                      gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)
```

used for generating random, non-critical data, can ignore

```
plugins/inputs/powerdns_recursor/protocol_v1.go:22:18  gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)
plugins/inputs/powerdns_recursor/protocol_v2.go:20:18  gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)
```

used to set up socket file with random-ish name, can ignore

```
plugins/inputs/statsd/running_stats.go:84:11           gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)
plugins/inputs/vsphere/endpoint.go:572:13           gosec  G404: Use of weak random number generator (math/rand instead of crypto/rand)   
```

both pick random index, can ignore

**Comment by zak-pawel** [comment]
@powersj Do you propose we should not enable this rule at all or we should enable it and ignore all tests occurences (in `.golangci.yml`) and explicitly ignore all existing occurences in non-test code?

**Comment by powersj** [comment]
I'd like to hear from @srebhan if he agrees with my assessment above first. However, I am leaning towards not enabling.

**Review by Hipska — APPROVED** [review]

**Comment by zak-pawel** [comment]
@srebhan A penny for your thoughts :)

**Comment by srebhan** [comment]
> @srebhan A penny for your thoughts :)

@zak-pawel usually I'm paid in beer... ;-P

In my opinion we should enable G404 to make us aware about weak RNGs during review. Otherwise I would silence the following occurrences without any code change:
- `internal/internal.go:128` and `internal/internal.go:146` as not security critical
- `plugins/inputs/mock/mock.go:69` as not security critical
- `plugins/inputs/statsd/running_stats.go:84` as not security critical
- `plugins/inputs/vsphere/endpoint.go:572`  as not security critical

For the following occurrences I suggest modifying the code as follows:
- `plugins/inputs/example/example.go:106` change the example to not use random numbers
- `plugins/inputs/powerdns_recursor/protocol_v1.go:22`  and ``plugins/inputs/powerdns_recursor/protocol_v2.go:20`` replace the random-number with either a `os.CreateTemp` construct or a UUID or omit the `laddr` and get the `LocalAddr` of the socket (not sure if this works in that case).

Does that help?

File: plugins/inputs/example/example.go [hunk 1]
```diff
@@ -101,11 +103,15 @@ func (m *Example) Gather(acc telegraf.Accumulator) error {
...
 		for i := int64(1); i < m.NumberFields; i++ {
 			name := fmt.Sprintf("field%d", i)
-			value := 0.0
+			var err error
+			value := big.NewInt(0)
 			if m.EnableRandomVariable {
-				value = rand.Float64()
+				value, err = rand.Int(rand.Reader, big.NewInt(math.MaxUint32))
+				if err != nil {
+					return err
```
Comment (Hipska) [inline plugins/inputs/example/example.go:111]: Aren’t you saying above to accumulate errors?

Comment (zak-pawel) [reply plugins/inputs/example/example.go]: @Hipska 
Good catch, changed.

**Review by Hipska — APPROVED** [review]

**Review by srebhan — APPROVED** [review]
Nice. Thanks @zak-pawel!

**Review by powersj — APPROVED** [review]

**Pull request merged**